### ISSUE TITLE
Обработка ошибок по человекочитаемому тексту

### ISSUE THREAD

**Issue by antonkalinin-ml** [unam3/fsd-rest-news#31]
В ProcessRequest некоторые ошибки определяются разбором человекочитаемого текста из постгреса:

https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/src/RestNews/DB/ProcessRequest.hs#L400
https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/src/RestNews/DB/ProcessRequest.hs#L572
https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/src/RestNews/DB/ProcessRequest.hs#L433

Человекочитаемые тексты не предназначены для машинного разбора, они могут отличаться в разных версиях постгреса или даже в одной версии - постгрес не гарантирует, что для определенного SQLSTATE текст ошибки всегда соответствует одному шаблону. Опираться можно только на код ошибки SQLSTATE и еще, возможно, на поля вроде [PG_DIAG_TABLE_NAME](https://www.postgresql.org/docs/13/libpq-exec.html). Проблема с PG_DIAG_TABLE_NAME в том, что hasql их не передает, и постгрес не гарантирует их наличие, и я никогда ими не пользовался и не уверен, что оно обозначает таблицу-цель foreign key constraint, а не таблицу, где произошла ошибка. Возможно, понадобится сделать проверочные запросы.

В RestNews одна ошибка тоже проверяется по текстовому сообщению:
https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/src/RestNews.hs#L225
Для человека "no such category" и "category not found" - одно и то же, а для обработчика, который ожидает фиксированную строку - нет. В других языках можно заводить именованные строковые константы, но и там предпочитают использовать однозначные константы-коды ошибок вида `"NO_SUCH_CATEGORY"`, а не прямо сообщение для юзера. ADT в хаскеле здесь будут еще удобнее.

**Comment by unam3** [comment]
> Человекочитаемые тексты не предназначены для машинного разбора

https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/src/RestNews/DB/Errors.hs#L47-L63

**Comment by antonkalinin-ml** [comment]
Да, это я видел. Я про другое - про код вида `Just "Key (tag_id)" -> ...`, где матчится тект ошибки. Здесь мы сперва определили по коду, что это foreign key constraint violation (согласен, что так можно), но какой именно констрейнт нарушен, определяется по строке ошибки.

**Comment by unam3** [comment]
Необходимо ли с этим ишью что-то делать при наличии вот этих строк?

https://github.com/unam3/fsd-rest-news/blob/325121ce052d8162e52b99286c15b2dfd1c6ba9c/README.md?plain=1#L7-L8

**Comment by antonkalinin-ml** [comment]
Исправь, пожалуйста, хотя бы в паре мест, если все места править трудоемко. В реальном проекте очень вероятно, что рано или поздно придется обновлять постгрес. У него цикл поддержки 5 лет, многие проекты живут дольше. Можно сперва взять в работу другие замечания, которые можно закрыть быстрее.
